Serialize FileIO and TextIOWrapper and Universe (#2723)
MDAnalysis/mdanalysis: package/doc/sphinx/source/documentation_pages/pickle_universe.rst

> .. Contains the formatted docstrings for the serialization of universe located 

Reviewer: Put this as under [doc/sphinx/source/documentation_pages/coordinates](https://github.com/MDAnalysis/mdanalysis/tree/develop/package/doc/sphinx/source/documentation_pages/coordinates) and add an entry to the second TOC (for developers) in [doc/sphinx/source/documentation_pages/coordinates_modules.rst](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/doc/sphinx/source/documentation_pages/coordinates_modules.rst):
```reST
The remaining pages are primarily of interest to developers. Programmers and anyone trying to implement new functionality should first read the :ref:`Trajectory API` and then look at the following pages. 

.. toctree::
   :maxdepth: 1

   coordinates/base
   coordinates/core
   coordinates/pickle_universe.rst
   coordinates/chain
   coordinates/XDR

In particular, all trajectory readers have to be :ref:`serializable<serialization>` and they should pass all tests
available in the ``MDAnalysisTests.coordinates.base.MultiframeReaderTest`` or ``MDAnalysisTests.coordinates.base.BaseReaderTest``.
```

.. mainly in 'MDAnalysis/libs/pickle_file_io.py'
.. _serialization:

*********************************************************
Serialization of Universes
*********************************************************

> .. module:: MDAnalysis.libs.picklable_file_io

Reviewer: This can't be here as it will compete with the actual module. Remove.

Reviewer: (And sphinx should flag it as an error to have the same module twice.)

file handle are saved. On unpickling, the file is opened by filename.
This means that for a successful unpickle, the original file still has to
be accessible with its filename. To retain the current frame of the trajectory,
:func:`_read_frame(previous frame)` will be called during unpickling.

Here is an example of how a :class:`Universe` is (un)pickled:

.. code-block:: python

    u = MDAnalysis.Universe(topology, trajectory)
    u.trajectory[5]
    u_pickled = pickle.loads(pickle.dumps(u)
    u_pickled.ts.frame == u.trajectory.ts.frame

.. _simple_parallel_analysis:

Simple analysis case
--------------------

.. code-block:: python
    def cog(u, ag, frame_id):
        u.trajectory[frame_id]
        return ag.center_of_geometry()

    u = MDAnalysis.Universe(GRO, XTC)
    ag = u.atoms[2:5]

    p = multiprocessing.Pool(2)
    result = np.array([p.apply(cog, args=(u, ag, i))
                       for i in range(u.n_frames)])
    p.close()

Note to make sure :class:`Atomgroup` finds its anchored :class:`Universe` after pickling,
the :class:`Universe` has to be first pickled.

.. _how_to_serialize_a_new_reader:

How to serialize a new reader
-----------------------------

> - File Access

Reviewer: heading, not lone bullet
```suggestion
File Access
~~~~~~~~~~~
```

If the new reader uses :func:`util.anyopen()` (e.g. PDB), the reading handler
can be pickled without modification.
If the new reader uses I/O classes from other package (e.g. GSD), and cannot
be pickled natively, create a new picklable class inherited from 
the file class in that package (e.g. GSDPicklable), adding :func:`__getstate__`,
:func:`__setstate__` functions to allow file handler serialization.

> - To seek or not to seek

Reviewer: subheading

Some I/O class supports :func:`seek` and :func:`tell` functions to allow the file 
to be pickled with an offset. It is normally not needed for MDAnalysis with
random access. But if error occurs, find a way to make the offset work.

> - Miscellaneous

Reviewer: subheading

If pickle still fails due to some unpicklable attributes, try to find a way

> to pickle those, or write custom :func:`__get/setstate__` methods for the reader.

Reviewer: only use markup for real functions, don't use a confusing mixed name
```suggestion
to pickle those, or write custom :func:`__getstate__` and :func:`__setstate__` methods for the reader.
```

If the new reader is written in Cython, read :class:`lib.formats.libmdaxdr` and
:class:`lib.formats.libdcd` files as references.

> - Tests

Reviewer: subheading

If the test for the new reader uses :class:`BaseReaderTest`, whether
the current timestep information is saved, and whether its relative
position is maintained, i.e. next() read the right next timestep,
are already tested.

If the new reader accesses the file with :func:`util.anyopen`, add necessary

> testes inside `parallelism/test_multiprocessing.py` for the reader.

Reviewer: two backtics for standard reST markup
```suggestion
testes inside ``parallelism/test_multiprocessing.py`` for the reader.
```

If the new reader accessed the file with new picklable I/O class,
add necessary tests inside `utils/test_pickleio.py` for I/O class,
`parallelism/test_multiprocessing.py` for the reader.

.. _implemented-fileio:

Currently implemented picklable FileIO Formats
----------------------------------------------

    :class:`MDAnalysis.lib.picklable_file_io.FileIOPicklable`
    :class:`MDAnalysis.lib.picklable_file_io.BufferIOPicklable`
    :class:`MDAnalysis.lib.picklable_file_io.TextIOPicklable`
    :class:`MDAnalysis.lib.picklable_file_io.BZ2Picklable`
    :class:`MDAnalysis.lib.picklable_file_io.GzipPicklable`
    :class:`MDAnalysis.coordinates.GSD.GSDPicklable`
    :class:`MDAnalysis.coordinates.TRJ.NCDFPicklable`

>     :class:`MDAnalysis.coordinates.chemfiles.ChemfilesPicklable`

Reviewer: Make it a list with bulletpoints.
